DNM: Export; WIP

miretskiy opened the pull request:
See Commits.

Work in progress.  Do not merge.

Commit:
jobs,schedule: Refactor schedule options handling into `schedulebase` package.

Move and refactor code related to parsing schedule specific options, such
as `first_run`, `on_previous_running` and `on_error` into a standalone
`schedulebase` package so that that code maybe reused by new schedule
implementations.

Release Notes: None

Commit:
changefeedccl: Implement `EXPORT` mode for changefeeds.

Add ability to execute changefeeds in "export" mode:

```
CREATE EXPORT CHANGEFEED FOR TABLE tbl
[AS OF SYSTEM TIME ...] INTO 'sink://...' [WITH opt=value,....]
```

When changefeed is created with `EXPORT` mode, the changefeed
will only execute the backfill portion of changefeed: that is,
it will emit all current values in the table into the specified
sink.

In addition, it is possible to specify an `AS OF SYSTEM TIME` clause
to when creating export changefeeds so that only the data as of the
specified time is emitted.

The export mode changefeed for table `tbl' is somewhat similar to
`EXPORT INTO CSV 'destination://' FROM SELECT * FROM tbl`.
There are significant differences though:
  * The changefeed export is executed as a job -- and as such, it can
    checkpoint its progress, thus being able to deal with much larger
    scale export.
  * Any format and sink destination supported by changefeed are supported
    (avro format, JSON, export to kafka, etc)
  * Any limitations present in the current changefeed system are applicable
    to this type of export.  In particular, it is possible that some
    duplicate records will be emitted, and we currently support export
    for a single column family tables.

Release Notes (enterprise): Make it possible to execute enterprise
changefeeds in export mode: that is, the entirety of the table is emitted
as a changefeed stream, after which the changefeed shuts down.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/70057)
<!-- Reviewable:end -->

Commit:
changefeedccl: Implement scheduled job for table export.

Implement a `CREATE SCHEDULE FOR EXPORT TABLE` statement.

The schedule export creates a scheduled job responsible for
execution of the periodic exports.

Release Justification: DO NOT RELEASE.  Just a placeholder
to keep linter happy.

Release Note (enterprise): Implement `CREATE SCHEDULE FOR EXPORT` statement
responsible for periodic export execution.

miretskiy commented:
@stevendanna @HonoreDB : Don't know if I'll wind up merging this changefeed export stuff; but would appreciate at least some cursory review at this point.

HonoreDB reviewed (commented):
> Reviewed 28 of 42 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70057)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @shermanCRL, and @stevendanna)

___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70057#-Mk3wzCn8Ze1ebrCCYmq:-Mk3wzCo9WxPaJxlMCLS:bvp9q5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/e603afc6bc896e7675809aff3ef150ab0a93585b/pkg/ccl/changefeedccl/changefeed_processors.go#L1187)):*
> ```Go
> 				cf.frontier.boundaryType == jobspb.ResolvedSpan_RESTART) {
> 			var err error
> 			if !cf.spec.Feed.ExportMode {
> ```

I'd think we probably should error or restart if we manage to hit a schemaChangeBoundary during export mode; don't want to export an incomplete table or apply an inconsistent schema. Edit: Oh, you do handle this later, maybe just need a comment here.


<!-- Sent from Reviewable.io -->

shermanCRL commented:
I recall we (@dt) had ideas for dealing with duplicates here. Something to do with using cloud storage. Remind me?
